src/dmd/statementsem.d

```diff
@@ -2147,7 +2155,19 @@ else
         sc.callSuper = cs0;
         sc.fieldinit = fi0;
         if (ifs.elsebody)
-            ifs.elsebody = ifs.elsebody.semanticScope(sc, null, null);
+        {
+            Scope *sce = sc.push();
+            // Recognize if (!__ctfe) as a special form condition and allow allocations
+            // in @nogc functions if we're in a __ctfe branch.
+            if (ifs.condition.op == TOKnot && (cast(UnaExp)ifs.condition).e1.op == TOKvar)
+            {
+                VarExp ve = cast(VarExp)(cast(UnaExp)ifs.condition).e1;
+                if (ve.var && ve.var.ident == Id.ctfe)
```

Comment:
I'm seeing an opportunity to centralize some of this logic into a function.  It's already used twice in this function scope and also at https://github.com/dlang/dmd/blob/master/src/dmd/expressionsem.d#L7299.  Perhaps a `VarExp.is__ctfe()` member function?

Reply:
Maybe would be better in the if condition ast, and return a tristate to represent: no, yes if, yes else.

Such a function may benefit from caching also.  But is not necessary.